Introduces metabolites, tweaks basic medicine and medkits.

Rukofamicom opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

###  Metabolites: 
* Adds metabolites, a new type of reagent used as a flag for reagents that have been recently been processed by a given carbon. These can be utilized for a multitude of functions beyond what I have done in this PR, such as toxins leaving traces behind after being processed. 
* Implements metabolites for Bicaridine, Kelotane, Antitoxin, Tricordrazine, Styptic Powder and Silver Sulfadiazine.
* Changes the path on BZ metabolites and shifts it into the metabolites.dm
  * Removes the flags from BZ metabolites to bring it more closely in line with other metabolites - only produced as a byproduct within the body. 

### Basic reagents:
* Bicaridine, Kelotane, Anti-toxin and Tricordrazine now have their efficacy scaled based on the current quantity of metabolites in the user's system. The more metabolites someone has, the less effective they become. This means they can be used to treat minor to medium injuries, but if used on major injuries they can only do so much before they lose most of their efficacy.
* Bicaridine, Kelotane and Anti-toxin have had their overdoses tweaked - they now rapidly convert into metabolites and do half as much liver damage per tick.
* Tricordrazine overdose was reduced from 50 to 30 units and now does half as much liver damage per tick (but retains its toxin damage)
* Entirely removes the restrictions placed on Dexalin so that it functions at full power (1.5 oxy) regardless of patient condition and other factors. 

Here is a spreadsheet plotting out the exact values behind these changes:
https://docs.google.com/spreadsheets/d/111vmAeDzTc7QYerfL5hASpuF9Byta5Wcb3nyJyN-tzQ/

### White medkit items:
* Bruise packs and ointment now utilize reagent code to heal instead of magic healing powers with no regard for the reagents they contain and are implied to utilize.
* In order to reduce potential overdoses, each application now heals for half as much, but there are also twice as many applications per stack to keep the overall healing at the same level. This also means crafting them costs half as many reagents.
* Building on an idea @PowerfulBacon introduced in #8495, but with a bit more flavor behind it - players may not apply bruise packs or ointment while they currently have metabolites from either of the very painful medicines *"That stuff really hurt! You can't bring yourself to apply [src] to your wounds again yet, maybe someone else can help put it on for you."*.
* Gauze can now be used to treat simplemobs just like bruise packs.


<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

### Metabolites:
Are a good groundwork to lay for various features and ideas I've seen. There's been suggestions that medical and/or detective scanners should only be able to detect metabolites instead of all reagents, or effects that could be added to more gasses than just BZ. This PR is just one example of how they can be utilized moving forward - to restrict single-reagent spam.

### Basic chems:
* #7936 went way too far in its nerfs. The reagents aren't just for minor injuries injuries now, they border on being unusable and this impacts medical cyborgs especially hard given that these basic reagents and tend wounds are the only thing they can actually do without upgrades.
* A drop-off in efficacy based on how much of a given reagent has already been consumed ensures the reagent remains an accessible option in any situation, while ensuring it isn't the only reagent anyone will ever need for for even grave injuries. Current numbers allow for 10u to gradually heal 47 damage over almost two minutes - another 10u heals only an additional 23 over the same two minutes if metabolites aren't given time to disperse. *Patients in crit will need more than just a basic reagent to recover fully, but can still benefit from having been treated.*
* Tricordrazine keeps is dramatically increased metabolization rate so that it isn't an easy-access cure-all. It serves as a quick but short lived burst of every type of healing and will make a good addition to most mixes in limited quantities, but will never be good enough to heal most injuries on its own due to rapid metabolite buildup.
* Dexalin is something people reach for when a patient is already unconscious and by itself is barely able to do anything to help even at this full power as it is. It fulfilled the role of primarily being a "helper" reagent rather than a primary medicine as is.

### White medkit items:
* They're just too damn good. Being able to bring someone from the brink of death back to full health with almost no downside in a very short amount of time is not something that should be possible for the most basic and publicly accessible medkit.
* I agree with the premise of #8495 that the medkit items should be primarily used to treat others who are injured rather than oneself and I think this is a superior way to limit self-use in terms of both flavor and player restriction.
* Making the items rely on and utilize reagent code is just much more consistent design overall, especially when they very clearly contain/utilize a specific reagent. 

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

### Metabolites testing

![image](https://user-images.githubusercontent.com/9547572/221317755-c78285fb-f804-4dcb-be32-3c97ca542ebe.png)

![image](https://user-images.githubusercontent.com/9547572/221318042-e16dfaeb-cfe5-4f21-a6f1-3639a38fb363.png)

![image](https://user-images.githubusercontent.com/9547572/221318906-43f4b613-a5a1-4761-85cf-3270201f1e4f.png)

Capping at 15 metabolites:
![image](https://user-images.githubusercontent.com/9547572/221330601-24d3e3b3-1dd0-4a53-9763-241e20df4e55.png)


### Basic Reagents

Heal testing, Bicaridine, Kelotane, Antitoxin:
5u heals 25 damage over 25 ticks
10u heals 44 damage over 50 ticks
15u heals 56 damage over 75 ticks
20u heals 64 damage over 100 ticks
25u heals 70 damage over 125 ticks

Heal testing, Tricordrazine:
5u heals 10 each damage over 5 ticks
10u heals 15 each damage over 9 ticks
15u heals 19 each damage over 13 ticks
20u heals 21 each damage over 17 ticks
25u heals 23 each damage over 21 ticks

Overdose testing:
![image](https://user-images.githubusercontent.com/9547572/221319061-b062af96-4619-43af-b0fd-4d93050a5093.png)

33u (mild overdose) did this
![image](https://user-images.githubusercontent.com/9547572/221319152-9cbb84eb-6913-4887-8bbc-295c2225676f.png)

### White Medkit items:

Applying pack
Error for trying to heal something that's fine
Unable to apply second patch after trying because of pain
![image](https://user-images.githubusercontent.com/9547572/221329530-ca1ec37d-e71e-4e7e-9720-c189659f826e.png)

Trying to apply ointment to a bruise:
![image](https://user-images.githubusercontent.com/9547572/221329563-19998105-8c4d-4646-8190-96a300d59155.png)

Metabolites and reagents handled correctly:
![image](https://user-images.githubusercontent.com/9547572/221329787-4708c944-4453-4925-8972-9a279633f76f.png)
![image](https://user-images.githubusercontent.com/9547572/221329935-dabc7736-6fc1-4f2e-9b48-eb818ab36cb1.png)


Grinding handled correctly:
![image](https://user-images.githubusercontent.com/9547572/221329861-f96241cd-dbf0-4180-9463-b5d83a33bed1.png)
![image](https://user-images.githubusercontent.com/9547572/221329902-31b21fe9-a3d3-4d37-a3d1-040aad6102a7.png)
![image](https://user-images.githubusercontent.com/9547572/221329967-98115acf-51f8-4ffa-934c-1153a8da3e6c.png)

Able to apply another bruise pack/ointment after 50 reagent ticks (100 seconds), barring outside factors.

Icons also update as expected, thanks to a clever formula that was already present:
![image](https://user-images.githubusercontent.com/9547572/221330009-752fec3e-8ce1-474e-a440-052bcf739855.png)
![image](https://user-images.githubusercontent.com/9547572/221330024-56f1899d-eeb5-459b-895c-1195a86c2b78.png)
![image](https://user-images.githubusercontent.com/9547572/221330028-a2e413cc-32e9-450b-a065-94a70d89f00a.png)

And finally, gauze on simplemobs:
![image](https://user-images.githubusercontent.com/9547572/221330517-85215b68-2bb2-40db-bd5b-0a2759502d4f.png)

</details>

## Changelog
:cl:
add:  Adds a new type of reagent, metabolites. Metabolites are produced as a byproduct of any reagent with a metabolite set and can be used to trace what kinds of reagents someone has been exposed to at least somewhat recently. For now metabolites are only implemented for Bicaridine, Kelotane, Antitoxin, Tricordrazine, Styptic Powder and Silver Sulfadiazine.
balance: Bicaridine, Kelotane, Antitoxin and Tricordrazine are equally effective regardless of current HP again, instead having their effects wane as metabolites build up. They will always be useful, but will never be the only thing you need to treat more severe injuries.
tweak: Bruise packs and ointment now utilize reagent code to heal instead of directly healing. Due to overdosing concerns, they heal for half as much per application, but come with twice as many applications per stack to compensate.
balance: Because they now properly utilize medicine, ointments and bruise packs are now very painful! Players will find they are unable to subject themselves to this pain if they try to apply more than one in quick succession. Seek help from others if you need multiple applications!
tweak: Gauze can now heal injured simplemobs just like bruise packs.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Rukofamicom committed:
Adds metabolites

Rukofamicom committed:
Update metabolites.dm

Rukofamicom committed:
Update metabolites.dm

Rukofamicom committed:
Implements metabolites

Metabolite limitations replace overdose and other limiting factors for Bicaridine, Kelotane, Antitoxin and Tricordrazine.

Rukofamicom committed:
re-implements overdose

Extra metabolites as the body self-purges the excess
Liver takes some damage, but not enough to completely incapacitate it with a mild overdose

Rukofamicom committed:
Bruise Packs and Ointment

Now utilize reagent code and apply styptic powder and silver sulfadiazine respectively.

Grinding works as intended and expected.

Bruise packs and Gauze are still usable on simplemobs to heal their damage.

Rukofamicom committed:
more metabolites and new define

Styptic Powder
Silver Sulfadiazine

Definition for maximum metabolites to future-proof against potentially clogging a mob's reagent cap entirely with metabolites

Rukofamicom committed:
Prevent rapid application of bruise packs and ointment

Rukofamicom committed:
syntax

Rukofamicom committed:
missed define

Rukofamicom committed:
doubles quantity of med items

Rukofamicom committed:
tweak to error message

Rukofamicom committed:
bz metabolites moved to metabolites.dm

Rukofamicom committed:
Update medicine_reagents.dm

Rukofamicom committed:
Update medical.dm

Rukofamicom committed:
last minute tweaks

Rukofamicom committed:
Improved metabolite cap

EvilDragonfiend commented on code/modules/reagents/chemistry/reagents/medicine_reagents.dm:
```diff
@@ -1134,24 +1136,18 @@
 
 /datum/reagent/medicine/bicaridine/overdose_process(mob/living/M)
-	M.adjustOrganLoss(ORGAN_SLOT_LIVER, 2)
+	M.reagents.add_reagent(metabolite, 1)
```
You're doing this without using your own metabolite calculation.

Rukofamicom replied:
Yes, that's correct. This also isn't normal metabolite production. It's the body rushing to process an excess of chems and rapidly convert them into metabolites.

EvilDragonfiend replied:
well, I had to start a review not putting a single comment here by my mistake, but I got your answer.

EvilDragonfiend commented on code/modules/reagents/chemistry/reagents/medicine_reagents.dm:
```diff
@@ -15,6 +13,8 @@
 	current_cycle++
 	holder.remove_reagent(type, metabolization_rate / M.metabolism_efficiency) //medicine reagents stay longer if you have a better metabolism
+	if(metabolite)
+		holder.add_reagent(metabolite, metabolization_rate / M.metabolism_efficiency * METABOLITE_RATE)
```
make another proc that does `holder.add_reagent(metabolite, metabolization_rate / M.metabolism_efficiency * METABOLITE_RATE)`, and call it from here
and then let the bicaridine(others too) overdoes code call that proc too.

Rukofamicom replied:
That doesn't make sense though. 

Only three reagents are intended to have this as an overdose effect (at this time, at least) and they are not intended to follow the normal metabolite logic because they aren't being metabolized normally. Look at the handful of toxins that effectively produce histamine as a metabolite - the body views the reagent as foreign and is rushing to convert it. 

I do not want the overdose mechanic to match either of the two different normal processing rates. (one specific to medicine, and the one used for all other reagents)

EvilDragonfiend replied:
maybe not necessary since you wanted the overdose part working separated

EvilDragonfiend commented on code/modules/reagents/chemistry/reagents.dm:
```diff
@@ -1,4 +1,7 @@
 #define REM REAGENTS_EFFECT_MULTIPLIER
+#define METABOLITE_RATE     0.5 // How much of a reagent is converted metabolites if one is defined
+#define MAX_METABOLITES		15  // The maximum amount of a given metabolite someone can have at a time
+#define METABOLITE_PENALTY(path) min(max(M.reagents.get_reagent_amount(path)/2.5, 1), 5) //Ranges from 1 to 5 depending on level of metabolites. 
```
```suggestion
#define METABOLITE_PENALTY(path) clamp(M.reagents.get_reagent_amount(path)/2.5, 1, 5) //Ranges from 1 to 5 depending on level of metabolites. 
```

clamp() proc is identical to the mix of min(max())

> This is effectively equivalent to min(max(Number, Low), High), but with some slight differences. For one thing, clamp() is forgiving if you accidentally swap Low and High; it will just swap them back so Low is always lower. Also, because this is a single proc call it's slightly faster.
~DM docu

Rukofamicom replied:
Ah, thanks much.

EvilDragonfiend commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -85,12 +95,21 @@
-	if(affecting.heal_damage(heal_brute, heal_burn) || stop_bleeding)
+	if(reagent)
+		reagents.reaction(M, PATCH, affecting = affecting)
+		M.reagents.add_reagent_list(reagent) //Stack size is reduced by one instead of actually removing reagents from the stack.
```
there should be a code that reagent in a pack is removed by applied amount, so that you won't get the reagents from grinding those after using those.
so, `remove_reagent(medicine, REAGENT_AMOUNT_PER_ITEM/amount)` would work.

Rukofamicom replied:
Actually, this is how I initially did it and even had it working appropriately. I changed it to the system I'm using now because it took far less code/steps to achieve the desired results: 

It introduced a plethora of bugs when it comes to splitting stacks of bruise packs and ointments: When you split a stack an entirely new object is initialized, and reagents are not split. 

Say you have a stack of six ointment, which would be 120u of ointment. You split that into a separate stacks of 3 ointment each? You still have 120 ointment in the first stack, and have now initialized a new stack of ointment which will be given whatever the initial value of reagents was. There were two ways to solve this, and both were way more convoluted and introduced much worse snowflakey code than the solution I went with I feel like. 

In addition to the above: when trying to transfer reagents an entire stack of five is treated as one object when it comes to transferring reagents, so I'd be reducing the stack size by 1 *and also* manually reducing one stack's worth of reagents from the stack to ensure the stack maintained the correct amount of reagents.

There's so many ways for this to go wrong and bug out that I probably didn't encounter when I got it working too.

EvilDragonfiend commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -85,12 +95,21 @@
+	use(1)
+
+/obj/item/stack/medical/on_grind()
+	reagents.clear_reagents() //By default grinding returns all contained reagents + grind_results, and for stackable items we only want grind_results
```
yeah, but I think these should not have grind_results, and if you have manual reagent removal stuff, it's not necessary here.

Rukofamicom replied:
See above - it's a valid opinion and the code is slightly snowflakey to handle it this way, but the two alternatives were substantially worse in my opinion.

EvilDragonfiend commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -1,24 +1,34 @@
+/obj/item/stack/medical/Initialize(mapload, new_amount, merge, mob/user)
+	. = ..()
+	if(reagent)
+		create_reagents(REAGENT_AMOUNT_PER_ITEM)
```
I think it's fine to be 40u if you accept the change I suggested for how reagent works

Rukofamicom replied:
That would be 40u divided between the whole stack though wouldn't it? That would be an extraordinary nerf over the ability to be treated with a first aid kit.

EvilDragonfiend commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -1,24 +1,34 @@
+#define REAGENT_AMOUNT_PER_ITEM 20 //The amount of reagents medical items contain, for both application and grinding purposes.
```
this should be separated into
* REAGENT_AMOUNT_HEAL 20 // the amount of healing power
* REAGENT_AMOUNT_CONTAINED 40 // the amount of the medical items contain

Rukofamicom replied:
🤔 

Why on earth would I use two defines - defining how much is healed completely defeats the purpose of utilizing reagent code to determine how much is healed. The amount of reagents contained is **also** the healing efficacy because it transfers those reagents which then heals?

EvilDragonfiend commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -85,12 +95,21 @@
-	if(affecting.heal_damage(heal_brute, heal_burn) || stop_bleeding)
+	if(reagent)
+		reagents.reaction(M, PATCH, affecting = affecting)
+		M.reagents.add_reagent_list(reagent) //Stack size is reduced by one instead of actually removing reagents from the stack.
```
also, please don't use `M.reagents.add_reagent_list(reagent)` directly by taking `reagent` var.
reagent initialisation and reagent transfer to heal should be separated.

Rukofamicom replied:
As I've touched upon in my other responses: Stacks and reagents don't like to get along or play nice.

I'm using the initialized value because the values are effectively acting as flags for how much of what kind of reagent to transfer per use, rather than actually transferring the reagents due to all of the issues and extra snowflake code required for them to do so.

EvilDragonfiend reviewed (changes requested):
I am not sure how metabolite would work, so I will put no comment on those.

but bruise pack/ointment seems a bit logically weird in a part of how chem works. I'd like to point out these:

* If you created reagent in a pack, you need to transfer reagent to an applied mob. Since bruisepack/ointment's efficiency should be multiplied, manual add/removal instead of transfer will be needed.
* please separate the healing power and the reagent unit from the single define

It's a bit complex, so I am not sure how to explain it. so I will leave my review here anyway.

Rukofamicom committed:
Update code/modules/reagents/chemistry/reagents.dm

Co-authored-by: EvilDragonfiend <87972842+EvilDragonfiend@users.noreply.github.com>

Rukofamicom committed:
updates crafting recipes

One ointment/bruise pack should only take 20u reagents now, not 40

Rukofamicom committed:
Merge branch 'master' into medicine-basics

Rukofamicom committed:
fixes typo or something

Rukofamicom committed:
Tweaks message to user when trying to apply a second painful med item

That stuff really hurt! You can't bring yourself to apply [src] to your wounds again yet, maybe someone else can help put it on for you.

Changed to:

That stuff really hurt! You'll need to wait for the pain to go away before you can apply [src] to your wounds again, maybe someone else can help put it on for you.

PowerfulBacon reviewed (approved):
This seems okay and can be tweaked if needed as we go on to buff certain aspects if they are too weak.
